### PR TITLE
Added testing capabilities and completed the crypto implementations

### PR DESCRIPTION
**Pull request by oparkins**
## Overview
This pull request is to integrated the new testing capabilities and the crypto implementations into the master branch. This pull request includes 3 main issues: #9, #14, and #15. A side affect of the above issues is that the crypto methods (hmac and digest) were completed for both supported platforms. This made it easier to test all the changes made.

Also, warning, I am writing this late at night. I plan to look at this more tomorrow and see if it needs to be worked on more or not.

## Separating Classes (Issue #9)
@irremotus brought up the suggestion that the classes should be separate. To accomplish this without circular dependencies, I created a `cryptoFactory` class that return the proper crypto class to the caller. The factory will decide which class to return based on the availability of `crypto.subtle`.

The testing classes have also been separated and labeled different in Jest. The tests, described later, have been divided into two separate projects to mirror the separation of the crypto classes.

## Returning Promises (Issue #14)
To make the experience consistent among our supported platforms, all the crypto classes need to return promises. This is because the browser implementation of crypto requires promises and working around them is practically impossible. It is a cleaner design to wrap the sequential crypto calls from NodeJS in promises.

## Testing (Issue #15)
Testing is what took most of the time on this pull request. I took guidance from a [Blog](https://itnext.io/testing-your-javascript-in-a-browser-with-jest-puppeteer-express-and-webpack-c998a37ef887) online. There were quite a few changes made to accommodate the automatic testing of the browser code:
- Dockerfile
The Dockerfile was changed to install an instance of google-chrome headless. This is so that puppeteer (a npm tool) can control and test our browser code. 
- Admin privileges
The unfortunate side affect is that the docker container now requires the [CAP_SYS_ADMIN](https://lwn.net/Articles/486306/) capability to run properly.  This reduces the isolation of our tests from the host system, but it does allow us to test the code more thoroughly. 
- Puppeteer
Puppeeter is used in conjunction with Webpack and Jest to test the code. Puppeeter controls the headless chrome instance, Webpack packs up our code for browser use (removes require statements, etc), and then Jest runs the test code. There are a few new files just to support this capability. It may be possible to combine the files, but I did not look into that yet.
- procps
You will see this package in the Dockerfile. This was necessary to remove an error from Jest. Apparently Jest needs the program `ps` to run properly and this package provides it.

A new issue was discovered during this pull request, but it is a low priority issue. When testing the browser code, Webpack will load a lot of dependencies that are needed. It would be nice to remove that. This is documented in issue #16.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,12 +1,12 @@
 test: .build-tmp/build-image src/* test/*
-	docker run --rm -it -v ${PWD}/src:/app/src:ro -v ${PWD}/test:/app/test:ro rgpm:test
+	docker run --rm -it --cap-add=SYS_ADMIN -v ${PWD}/src:/app/src:ro -v ${PWD}/test:/app/test:ro rgpm:test
```
Comment (oparkins) [inline Makefile]: This is where the SYS_ADMIN privilege is added. It is required to run Chrome headless

File: test/browser/browserCrypto.test.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                return cryptoLib.toHex(await cryptoLib.digest(input));
+            }, input);
+            expect(result).toEqual(output);
+        });
+    });
+
+    describe("hmac method", () => {
+        it.each([
+            ["test", "test", "9ba1f63365a6caf66e46348f43cdef956015bea997adeb06e69007ee3ff517df10fc5eb860da3d43b82c2a040c931119d2dfc6d08e253742293a868cc2d82015"],
+            ["test", "test2", "8f3e4f699897645264cc0c9e6634fc9637929b932c931a4870e080ca3cfba0deb29a9573f676915ef701488ee3202d385ff03a306ce1333d5709491d73b08d62"],
```
Comment (oparkins) [inline test/browser/browserCrypto.test.js]: What do you think about these tests? Are they adequate enough you think?

File: test/server.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        <body>
+          <div id="root"></div>
+          ${paths.map((path) => `<script src="${path}"></script>`).join('')}
+        </body>
+      </html>`
+    )
+  })
+  .listen(port, () => {
+    console.log(`Server started at http://localhost:${port}/`)
+  })
```
Comment (irremotus) [inline test/server.js]:
```suggestion
  })

```

**Review by irremotus — CHANGES_REQUESTED** [review]

File: test/browser/browserCrypto.test.js (hunk 2 above)
Comment (oparkins) [reply test/browser/browserCrypto.test.js]: I've finally caught up on some work... I'll look into this today and commit something.

File: test/node/cryptoFactory.test.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        "digest",
+        "hmac",
+    ];
+
+    test.each(methodNames)("should exist", (methodName) => {
+        expect(crypto[methodName]).toBeDefined();
+    });
+
+
+    describe("digest method", () => {
```
[the remark this answers is not in the document]
Comment (oparkins) [reply test/node/cryptoFactory.test.js]: With how it is setup now, both subclasses that are created from cryptoFactory are tested. While it combines the cryptoFactory and generic crypto into one test file (which isn't my favorite), it tests this indirectly.

File: test/browser/browserCrypto.test.js (hunk 2 above)
Comment (oparkins) [reply test/browser/browserCrypto.test.js]: I think adding more tests can be an issue and can be a continuous process. I didn't add any new strings at the moment so we can focus on the structure of the tests

**Comment by oparkins** [comment]
Ignore the failing builds right now... There appears to be an issue with Travis-CI Xenial image. I put in a comment at https://travis-ci.community/t/xenial-kernel-does-not-appear-to-have-sandbox-capabilities/5370.

**Review by irremotus — CHANGES_REQUESTED** [review]
Looks great!  I unresolved one missed comment for a typo...

**Comment by oparkins** [comment]
Ignore the failing build. That is still due to the previous issue. It has been fixed in other branches. 

I fixed the typo for reals this time. @irremotus, can you approve the merge?

**Review by irremotus — APPROVED** [review]

**Pull request merged**